File: src/leetcode/python/medium/HouseRobberIII_337.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class HouseRobberIII_337(object):
    def rob(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """

        def dfs(node):
            if node is None:
                return [0, 0]
            left, right = dfs(node.left), dfs(node.right)
            return [node.val + left[1] + right[1],
                    max(left[0] + right[0], left[0] + right[1], left[1] + right[0], left[1] + right[1])]

        return max(dfs(root))

File: src/leetcode/java/HouseRobberIII_337.java
package leetcode.java;

public class HouseRobberIII_337 {
  public int rob(TreeNode root) {
    int res[] = robSubTree(root);
    return Math.max(res[0], res[1]);
  }

  private int[] robSubTree(TreeNode node) {
    if (node == null) return new int[2];
    int[] left = robSubTree(node.left);
    int[] right = robSubTree(node.right);
    int[] res = new int[2];
    res[0] = Math.max(left[0], left[1]) + Math.max(right[0], right[1]);
    res[1] = node.val + left[0] + right[0];
    return res;
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
      val = x;
    }
  }
}
